ext/msgpack/buffer.h

```diff
@@ -451,11 +451,45 @@ static inline VALUE msgpack_buffer_read_top_as_string(msgpack_buffer_t* b, size_
     }
 #endif
 
-    VALUE result = rb_str_new(b->read_buffer, length);
+    VALUE result;
+
+#ifdef HAVE_RB_ENC_INTERNED_STR
+    if (will_be_frozen) {
+        result = rb_enc_interned_str(b->read_buffer, length, utf8 ? rb_utf8_encoding() : rb_ascii8bit_encoding());
+    } else {
+        if (utf8) {
+            result = rb_utf8_str_new(b->read_buffer, length);
+        } else {
+            result = rb_str_new(b->read_buffer, length);
+        }
+    }
     _msgpack_buffer_consumed(b, length);
     return result;
-}
 
+#else
 
+    result = rb_str_new(b->read_buffer, length);
+#ifdef COMPAT_HAVE_ENCODING
```

Comment:
Not sure what the support matrix is, but if we could get rid of `COMPAT_HAVE_ENCODING` we could more easily directly instantiate the string with `rb_utf8_str_new` which would be preferable.

If this for Ruby 1.8?

Reply:
Yes. Ruby 1.8 is of course out of the support list of msgpack-ruby.
I strongly believe that we can remove `COMPAT_HAVE_ENCODING`, but it should be out of this pull request.

Reply:
I can try to do it if you'd like, we could then rebase this PR.

Reply:
That is definitely helpful to me :D